modules/ding_paragraphs/DingParagraphsBPIHelper.php

```diff
@@ -0,0 +1,365 @@
+<?php
+
+/**
+ * @file
+ * Code for the Ding paragraphs helper class.
+ */
+
+use Bpi\Sdk\Item\Node;
+
+/**
+ * Helper for paragraphs content in bpi.
+ *
+ * The public methods are called from BPI api hooks (cf. implementation of
+ * bpi.api.php in ding_paragraphs.module).
+ *
+ * The code looks complex in some places, but it is basically just runs
+ * through paragraphs structures to collect and serialize data to send to BPI,
+ * or runs through serialized data for rebuilding paragraphs when syndicating from BPI.
+ *
+ * Most of the code * inside the loops are similar to the code in the bpi
+ * module when pushing and syndicating content.
+ */
+class DingParagraphsBPIHelper {
+  const BPI_FILE_URL = 'bpi_file_url';
+  const DING_ENTITY_ID = 'ding_entity_id';
+
+  /**
+   * Process all files references from paragraphs.
+   *
+   * Stores the full file url in the data for later retrieval when syndicating
+   * paragraphs.
+   *
+   * @param array $paragraphs_data
+   *   The paragraphs data.
+   *
+   * @return array
+   *   The BPI assets.
+   */
+  public function replaceFilesWithUrls(array &$paragraphs_data) {
+    $files = [];
+
+    $this->traverse($paragraphs_data, function ($key, &$value, &$data) use (&$files) {
+      if (is_array($value) && isset($value['fid'], $value['uri'])) {
+        $file = file_load($value['fid']);
+        if ($file) {
+          $files[] = $file;
+          $value += array(self::BPI_FILE_URL => file_create_url($file->uri));
+        }
+      }
+    });
+
+    return $this->getBpiAssets($files);
+  }
+
+  /**
+   * Process material references in paragraphs data.
+   *
+   * Stores the actual ding entity id in the data.
+   *
+   * @param array $paragraphs_data
+   *   The paragraphs data.
+   */
+  public function processMaterials(array &$paragraphs_data) {
+    $this->traverse($paragraphs_data, function ($key, &$item) {
+      if (is_array($item) && !empty($item['ting_object_id'])) {
```

Comment:
@rimi-itk Do you have something to add here?

Reply:
This is lifted from https://github.com/ding2/ding2/blob/master/modules/bpi/bpi.push.inc#L159-L186 (apart from us using just `$item->endpoints` rather than `$item['value']->endpoints`). 

Extensive manual testing have been done and this works as expected.

Reply:
Please see the comment in the other discussion as there is nothing we can do about this. It's paragraph "magic".